Add filters to package cataloger

jonasagx opened the pull request:
This PR adds filters so a package without name or version doesn't go in the list of all discovered packages. Two catalogers were modified:
* generic cataloger: used by most specific catalogers, acts as a central-ish way of guaranteeing packages have a name.
* python package cataloger: doesn't use the generic cataloger, therefore needed its own package filter.

This PR addresses items 1, 2 and 3 from https://github.com/anchore/syft/issues/780#issuecomment-1142664048.

Integration and units tests were added to validate behavior.

Fix: #780 

Signed-off-by: Jonas Xavier <jonasx@anchore.com>

jonasagx commented:
SPDX doesn't require [package versions](https://spdx.github.io/spdx-spec/package-information/#73-package-version-field), while [CycloneDX does](https://cyclonedx.org/docs/1.2/json/#components_items_version) we should be compatible with both formats and only require package name.

stevespringett commented:
As of CDX v1.4, component version is optional, not required.

luhring reviewed (changes requested):
Thanks for getting this started! This isn't ready for merging just yet.

At the highest level, we need to figure out the best place to filter out unacceptable packages. As implemented currently, this filtering happens fairly downstream in the process. This is good in that there's a central place where filtering happens, but it's not good in that it means unusable packages are flowing through more parts of the logic. This makes the code a bit trickier to debug and makes it easier to miss problems — such as that we haven't yet accounted for package relationships properly (I commented on this below). So I think we need to figure out a better place to do this filtering.

I see that you've added CLI tests, but it's not yet clear to me why these tests needed to be created at the "CLI" level. I've asked about this in a code comment below.

The PR description says that you've added integration tests, but I don't see new integration tests. Is there more code to check into this branch? (I do see new test fixture files added.)

See the other code comments for the rest of my thoughts — although some may become moot depending on how the implementation ends up pivoting.

Let me know if you have any questions!

jonasagx commented:
Thank you for the review @luhring!

> At the highest level, we need to figure out the best place to filter out unacceptable packages. As implemented currently, this filtering happens fairly downstream in the process. This is good in that there's a central place where filtering happens, but it's not good in that it means unusable packages are flowing through more parts of the logic. This makes the code a bit trickier to debug and makes it easier to miss problems — such as that we haven't yet accounted for package relationships properly (I commented on this below). So I think we need to figure out a better place to do this filtering.

I added package filtering validation to the **generic cataloger**, used by most specific catalogers. I added code removing impacted relationships.

The Python Package Cataloger doesn't use the generic cataloger, I had to change it to filter out packages without name. This cataloger doesn't gather relationships so the change is pretty simple.
 
> The PR description says that you've added integration tests, but I don't see new integration tests. Is there more code to check into this branch? (I do see new test fixture files added.)

I removed the cli tests in favor of unit and integration tests. Before I had added only integration test _fixtures_, sorry for the confusion.

jonasagx commented on test/cli/package_missing_data_test.go:
```diff
@@ -0,0 +1,38 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFilterOutPackagesWithNoNameOrVersion(t *testing.T) {
```
I replaced this test with unit tests and integration test, resolving this comment.

jonasagx commented on syft/pkg/cataloger/python/package_cataloger.go:
```diff
@@ -51,13 +51,17 @@ func (c *PackageCataloger) Catalog(resolver source.FileResolver) ([]pkg.Package,
 			pkgs = append(pkgs, *p)
 		}
 	}
 	return pkgs, nil, nil
```
Notice this cataloger doesn't return relationships

jonasagx commented on syft/pkg/cataloger/catalog.go:
```diff
@@ -104,6 +109,21 @@ func Catalog(resolver source.FileResolver, release *linux.Release, catalogers ..
+
+func hasEmptyName(p pkg.Package) bool {
+	return p.Name == ""
+}
```
Correct, this code no longer exists. Generic and Python Package catalogers do the equivalent filtering.

jonasagx commented on syft/pkg/cataloger/catalog.go:
```diff
@@ -85,6 +86,10 @@ func Catalog(resolver source.FileResolver, release *linux.Release, catalogers ..
 
+			if filterOutPkg(p, hasEmptyName) {
+				continue
+			}
```
I addressed items 1 and 2 with the generic cataloger and the python package cataloger. In the generic cataloger I remove all relationships From and To a dropped package. 

> So this filtering probably needs to be done at least as upstream as in the catalogers themselves, so that we never generate relationships for packages we're going to discard.
Having package and relationships dropped by the generic cataloger still a (centralized) last barrier for invalid packages.

jonasagx commented on test/cli/test-fixtures/image-empty-files/pkgs/broken-2.0.0-py3.8.egg-info:
```diff
@@ -0,0 +1 @@
+Name: broken-2.0.0
```
Thanks for pointing that out. I moved it to the integration tests folder of fixtures, under a new file: `no-version`.

jonasagx commented:
Other than the `Python Package Cataloger` I've found a couple others that don't depend on generic cataloger:
* `deb.NewDpkgdbCataloger()` requires version and name to make a package 👍🏼 
* `rpmdb.NewRpmdbCataloger()` doesn't enforce non-empty name, but it depends on the data from an rpmDBs, I don't know how reliable they are 🤞🏼 
* `golang.NewGoModuleBinaryCataloger()` relies on go mod path, which is always present 👍🏼

luhring reviewed (changes requested):
This is getting closer... 👍 

I left some specific comments, but I'll also provide some high-level feedback here:

1. There's still an integration test being leveraged here, but again, we don't need an integration test here — we can just use a unit test. Using a unit test is preferable wherever possible.

2. We should have a central function that checks is a package is viable (i.e. has a name). A function like this is defined in the `python` package, but it should be defined instead in a central place, and then used by all the catalogers, so that we're applying the same criteria to packages from all catalogers. I saw your comment about some of the catalogers maybe being okay without this check, but I'd strongly encourage the use of this check for all code paths, so that we can centrally control what the requirements are for a viable package. It's not clear to me that the assumptions about other catalogers like RPM and Go are sufficient here.

jonasagx commented on syft/pkg/cataloger/common/generic_cataloger.go:
```diff
@@ -56,18 +56,47 @@ func (c *GenericCataloger) Catalog(resolver source.FileResolver) ([]pkg.Package,
 			p.SetID()
+			// doing it here so all packages have an ID,
+			// IDs are used to remove relationships
+			if p.Name == "" {
```
I moved the validation to a func in the `pkg` package (:P words...), that gets reused by other catalogers.

jonasagx commented on syft/pkg/cataloger/common/generic_cataloger.go:
```diff
@@ -56,18 +56,47 @@ func (c *GenericCataloger) Catalog(resolver source.FileResolver) ([]pkg.Package,
 	return packages, relationships, nil
 }
 
+func removePkgsFromRelationships(pkgs []artifact.Identifiable, relationships []artifact.Relationship) []artifact.Relationship {
```
Good call, I renamed the function.

jonasagx commented on syft/pkg/cataloger/common/generic_cataloger.go:
```diff
@@ -56,18 +56,47 @@ func (c *GenericCataloger) Catalog(resolver source.FileResolver) ([]pkg.Package,
+		}
+		cleanedRelationships = append(cleanedRelationships, r)
+	skip:
+	}
```
I agree, I implemented option 2.

jonasagx commented on syft/pkg/cataloger/python/package_cataloger.go:
```diff
@@ -51,13 +51,17 @@ func (c *PackageCataloger) Catalog(resolver source.FileResolver) ([]pkg.Package,
 	return pkgs, nil, nil
 }
 
+func pkgHasName(p *pkg.Package) bool {
```
The new pkg validation check for nil values as well.

jonasagx commented:
> I saw your comment about some of the catalogers maybe being okay without this check, but I'd strongly encourage the use of this check for all code paths, so that we can centrally control what the requirements are for a viable package. It's not clear to me that the assumptions about other catalogers like RPM and Go are sufficient here.

I've added pkg validation to deb/rpm/golang catalogers, I expanded the unit tests to ensure correct dropping of invalid packages.

luhring commented on syft/pkg/cataloger/common/generic_cataloger.go:
```diff
@@ -56,18 +56,44 @@ func (c *GenericCataloger) Catalog(resolver source.FileResolver) ([]pkg.Package,
 		}
 
-		relationships = append(relationships, discoveredRelationships...)
+		relationships = removeRelationshipsWithArtifactIDs(pkgsForRemoval, discoveredRelationships)
```
Is this logic correct? Since the `append` is removed, this is resetting the `relationships` value on each iteration of the `for` loop, losing prior iterations' relationships.

jonasagx replied:
That is a good bug. Thank you for pointing it out. I've fixed it and added a unit test to make it less likely to happen again.

luhring commented on syft/pkg/cataloger/common/generic_cataloger.go:
```diff
@@ -56,18 +56,44 @@ func (c *GenericCataloger) Catalog(resolver source.FileResolver) ([]pkg.Package,
 			continue
 		}
 
+		pkgsForRemoval := make(map[artifact.ID]artifact.Identifiable)
```
nit: Since you're never accessing the "value" in the key-value pair (i.e. the `artifact.Identifiable` value here), you don't need to store an `artifact.Identifiable` in the map. You can just use a map type of `map[artifact.ID]struct{}`, and store `struct{}{}` values to indicate that a given `artifact.ID` is present in the set. This uses less memory, but is also simpler to follow, since it makes clear that the map item's value is never used, other than to determine its existence.

jonasagx replied:
Interesting, I like that. I found [benchmark agreeing with you](https://gist.github.com/davecheney/3be245c92b61e5045f75).

luhring commented on syft/pkg/cataloger/common/generic_cataloger_test.go:
```diff
@@ -26,17 +26,16 @@ func parser(_ string, reader io.Reader) ([]*pkg.Package, []artifact.Relationship
-
 	globParsers := map[string]ParserFn{
-		"**/a-path.txt": parser,
+		"**/*.txt": parser, // this glob should capture all files, including the empty one
```
Can you explain this change? I want to understand your thinking here

jonasagx replied:
Yes, this glob (`**/*.txt`) will find a file named `empty.txt`, that will turn into an invalid package, and relationship, later on dropped from the final list. The previous glob (`**/a-path.txt`) wouldn't find `empty.txt`.

luhring reviewed (commented):
I think we're close... a couple of things to clear up

luhring reviewed (approved):
Looks great, thanks!
